prov/tcp/src/xnet_progress.c

```diff
@@ -386,12 +388,14 @@ static void xnet_progress_tx(struct xnet_ep *ep)
 	while (ep->cur_tx.entry) {
 		ret = xnet_send_msg(ep);
 		if (OFI_SOCK_TRY_SND_RCV_AGAIN(-ret)) {
-			xnet_update_pollflag(ep, POLLOUT, true);
-			return;
+			ret = xnet_update_pollflag(ep, POLLOUT, true);
+			if (!ret)
+				return;
 		} else if (ret == -OFI_EINPROGRESS_URING) {
-			xnet_update_pollflag(ep, POLLOUT, false);
-			return;
-		}
+			ret = xnet_update_pollflag(ep, POLLOUT, false);
+			if (!ret)
+				return;
+		 }
```

Comment:
My concern with the above change is that the handling of xnet_update_pollflag() isn't obvious in our intent.  For example, xnet_send_msg() could have been successful, with data sent on the wire.  However, if xnet_update_pollflag() fails, we fall through and fail the send (xnet_complete_tx() below).  AFAICT, the ep state remains connected.  But xnet_send_msg() could have written the header into the tcp stream.  If we drop the message, the stream is now corrupt, as the rest of the message won't be sent.

It may be better on xnet_update_pollflag() failure to jump to where xnet_ep_disable() is called.  That should fail the queued send as canceled.

Reply:
Thanks for the explanation. I have updated the code as you suggested and the function now disables the EP if `xnet_update_pollflag()` returns an error.